File: app/src/main/java/com/bluzi/babymonitor/net/Sockets.kt
package com.bluzi.babymonitor.net

import java.io.DataInputStream
import java.io.EOFException
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.Socket
import java.net.SocketTimeoutException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.withContext

// Socket seams for the CS2 transport — real impls here, scripted fakes in tests.

data class Datagram(val data: ByteArray, val host: String, val port: Int)

interface UdpSocket {
    suspend fun bind()
    suspend fun send(data: ByteArray, host: String, port: Int)
    suspend fun receive(): Datagram
    fun close()
}

interface TcpSocket {
    suspend fun connect(host: String, port: Int)
    suspend fun write(data: ByteArray)
    suspend fun readExact(n: Int): ByteArray
    fun close()
}

interface SocketFactory {
    fun udp(): UdpSocket
    fun tcp(): TcpSocket
}

object JavaSocketFactory : SocketFactory {
    override fun udp(): UdpSocket = JavaUdpSocket()
    override fun tcp(): TcpSocket = JavaTcpSocket()
}

private class JavaUdpSocket : UdpSocket {
    private var socket: DatagramSocket? = null

    override suspend fun bind() = withContext(Dispatchers.IO) {
        // WATCH-8: poll rather than block indefinitely. A raw DatagramSocket.receive() cannot be
        // interrupted by coroutine cancellation, so without this a camera that never answers would
        // park the handshake forever — no timeout could fire and no further attempt would be made.
        socket = DatagramSocket().apply { soTimeout = UDP_POLL_TIMEOUT_MS }
    }

    override suspend fun send(data: ByteArray, host: String, port: Int) = withContext(Dispatchers.IO) {
        val s = socket ?: throw IllegalStateException("udp: not bound")
        s.send(DatagramPacket(data, data.size, InetAddress.getByName(host), port))
    }

    override suspend fun receive(): Datagram = withContext(Dispatchers.IO) {
        val s = socket ?: throw IllegalStateException("udp: not bound")
        val buf = ByteArray(2048)
        while (true) {
            ensureActive() // the poll gap is where cancellation and withTimeout get their chance
            val pkt = DatagramPacket(buf, buf.size)
            try {
                s.receive(pkt)
                return@withContext Datagram(buf.copyOf(pkt.length), pkt.address.hostAddress ?: "", pkt.port)
            } catch (_: SocketTimeoutException) {
                // Nothing yet — go round again so we stay interruptible.
            }
        }
        @Suppress("UNREACHABLE_CODE")
        throw IllegalStateException("udp: unreachable")
    }

    override fun close() {
        socket?.close()
    }
}

private const val CONNECT_TIMEOUT_MS = 5_000
private const val READ_TIMEOUT_MS = 15_000
private const val UDP_POLL_TIMEOUT_MS = 500

private class JavaTcpSocket : TcpSocket {
    private var socket: Socket? = null
    private var input: DataInputStream? = null

    override suspend fun connect(host: String, port: Int) = withContext(Dispatchers.IO) {
        val s = Socket()
        s.tcpNoDelay = true
        s.connect(InetSocketAddress(host, port), CONNECT_TIMEOUT_MS)
        // WATCH-7: a camera that goes quiet must not hold the read open until TCP gives up (tens
        // of minutes). The camera pings us every second, so this much silence means it is gone.
        s.soTimeout = READ_TIMEOUT_MS
        input = DataInputStream(s.getInputStream())
        socket = s
    }

    override suspend fun write(data: ByteArray) = withContext(Dispatchers.IO) {
        val s = socket ?: throw IllegalStateException("tcp: not connected")
        s.getOutputStream().let {
            it.write(data)
            it.flush()
        }
    }

    override suspend fun readExact(n: Int): ByteArray = withContext(Dispatchers.IO) {
        val i = input ?: throw IllegalStateException("tcp: not connected")
        val buf = ByteArray(n)
        try {
            i.readFully(buf)
        } catch (e: EOFException) {
            throw XiaomiSocketClosed("tcp: connection closed", e)
        }
        buf
    }

    override fun close() {
        try {
            socket?.close()
        } catch (_: Exception) {
        }
    }
}

class XiaomiSocketClosed(message: String, cause: Throwable? = null) : Exception(message, cause)

File: windows/src/BabyMonitor.Core/Net/Sockets.cs
using System.Net;
using System.Net.Sockets;

namespace BabyMonitor.Core.Net;

/// <summary>
/// Socket seams for the CS2 transport — real sockets in the app, scripted fakes in the tests.
///
/// Two properties every real implementation must have, and both are tested against the real sockets
/// (SocketsTest), never against a fake:
///
///  - WATCH-8: a read must be interruptible. A camera that never answers must not park the handshake
///    forever — no timeout could fire, no further attempt would be made, and the app would sit on
///    "Connecting…" all night, having silently stopped looking.
///  - WATCH-7: a read must give up on its own. A camera that goes quiet must not hold the read open
///    until TCP's own timeout, which is tens of minutes away.
/// </summary>
public sealed record Datagram(byte[] Data, string Host, int Port);

public interface IUdpSocket : IDisposable
{
    Task BindAsync();

    Task SendAsync(byte[] data, string host, int port, CancellationToken ct = default);

    Task<Datagram> ReceiveAsync(CancellationToken ct = default);

    void Close();
}

public interface ITcpSocket : IDisposable
{
    Task ConnectAsync(string host, int port, CancellationToken ct = default);

    Task WriteAsync(byte[] data, CancellationToken ct = default);

    Task<byte[]> ReadExactAsync(int n, CancellationToken ct = default);

    void Close();
}

public interface ISocketFactory
{
    IUdpSocket Udp();

    ITcpSocket Tcp();
}

public sealed class SocketClosedException : Exception
{
    public SocketClosedException(string message, Exception? cause = null)
        : base(message, cause)
    {
    }
}

public static class SocketTimeouts
{
    public const int ConnectMs = 5_000;
    public const int ReadMs = 15_000;
}

public sealed class SystemSocketFactory : ISocketFactory
{
    public static readonly SystemSocketFactory Shared = new();

    public IUdpSocket Udp() => new SystemUdpSocket();

    public ITcpSocket Tcp() => new SystemTcpSocket();
}

internal sealed class SystemUdpSocket : IUdpSocket
{
    private readonly Socket _socket = new(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
    private readonly byte[] _buffer = new byte[65536];
    private bool _closed;

    public Task BindAsync()
    {
        _socket.Bind(new IPEndPoint(IPAddress.Any, 0));
        return Task.CompletedTask;
    }

    public async Task SendAsync(byte[] data, string host, int port, CancellationToken ct = default)
    {
        if (_closed)
        {
            throw new SocketClosedException("udp: socket closed");
        }

        var endpoint = new IPEndPoint(IPAddress.Parse(host), port);
        await _socket.SendToAsync(data, SocketFlags.None, endpoint, ct).ConfigureAwait(false);
    }

    /// <summary>
    /// WATCH-7/8: bounded and cancellable. A camera that stops answering fails the read rather than
    /// holding it open, and a cancellation (user stop, camera switch) takes effect at once.
    /// </summary>
    public async Task<Datagram> ReceiveAsync(CancellationToken ct = default)
    {
        using var timeout = CancellationTokenSource.CreateLinkedTokenSource(ct);
        timeout.CancelAfter(SocketTimeouts.ReadMs);
        try
        {
            var from = new IPEndPoint(IPAddress.Any, 0);
            var result = await _socket
                .ReceiveFromAsync(_buffer, SocketFlags.None, from, timeout.Token)
                .ConfigureAwait(false);
            var remote = (IPEndPoint)result.RemoteEndPoint;
            var data = new byte[result.ReceivedBytes];
            Buffer.BlockCopy(_buffer, 0, data, 0, result.ReceivedBytes);
            return new Datagram(data, remote.Address.ToString(), remote.Port);
        }
        catch (OperationCanceledException) when (!ct.IsCancellationRequested)
        {
            throw new SocketClosedException("udp: read timed out");
        }
        catch (ObjectDisposedException e)
        {
            throw new SocketClosedException("udp: socket closed", e);
        }
        catch (SocketException e)
        {
            throw new SocketClosedException($"udp: {e.Message}", e);
        }
    }

    public void Close()
    {
        if (_closed)
        {
            return;
        }

        _closed = true;
        try
        {
            _socket.Close();
        }
        catch (SocketException)
        {
            // Closing a socket that is already gone is not a failure worth telling anyone about.
        }
    }

    public void Dispose()
    {
        Close();
        _socket.Dispose();
    }
}

internal sealed class SystemTcpSocket : ITcpSocket
{
    private readonly Socket _socket = new(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
    private bool _closed;

    public async Task ConnectAsync(string host, int port, CancellationToken ct = default)
    {
        using var timeout = CancellationTokenSource.CreateLinkedTokenSource(ct);
        timeout.CancelAfter(SocketTimeouts.ConnectMs);
        try
        {
            _socket.NoDelay = true; // a monitor's frames are latency, not throughput
            await _socket.ConnectAsync(IPAddress.Parse(host), port, timeout.Token).ConfigureAwait(false);
        }
        catch (OperationCanceledException) when (!ct.IsCancellationRequested)
        {
            throw new SocketClosedException($"tcp: connect to {host}:{port} timed out");
        }
        catch (SocketException e)
        {
            throw new SocketClosedException($"tcp: connect to {host}:{port} failed — {e.Message}", e);
        }
    }

    public async Task WriteAsync(byte[] data, CancellationToken ct = default)
    {
        if (_closed)
        {
            throw new SocketClosedException("tcp: socket closed");
        }

        try
        {
            var sent = 0;
            while (sent < data.Length)
            {
                sent += await _socket
                    .SendAsync(data.AsMemory(sent, data.Length - sent), SocketFlags.None, ct)
                    .ConfigureAwait(false);
            }
        }
        catch (Exception e) when (e is SocketException or ObjectDisposedException)
        {
            throw new SocketClosedException($"tcp: write failed — {e.Message}", e);
        }
    }

    /// <summary>WATCH-7/8: bounded and cancellable, for the same reasons as the UDP read above.</summary>
    public async Task<byte[]> ReadExactAsync(int n, CancellationToken ct = default)
    {
        var out_ = new byte[n];
        var read = 0;
        while (read < n)
        {
            using var timeout = CancellationTokenSource.CreateLinkedTokenSource(ct);
            timeout.CancelAfter(SocketTimeouts.ReadMs);
            int got;
            try
            {
                got = await _socket
                    .ReceiveAsync(out_.AsMemory(read, n - read), SocketFlags.None, timeout.Token)
                    .ConfigureAwait(false);
            }
            catch (OperationCanceledException) when (!ct.IsCancellationRequested)
            {
                throw new SocketClosedException("tcp: read timed out");
            }
            catch (Exception e) when (e is SocketException or ObjectDisposedException)
            {
                throw new SocketClosedException($"tcp: read failed — {e.Message}", e);
            }

            if (got <= 0)
            {
                throw new SocketClosedException("tcp: connection closed by peer");
            }

            read += got;
        }

        return out_;
    }

    public void Close()
    {
        if (_closed)
        {
            return;
        }

        _closed = true;
        try
        {
            _socket.Close();
        }
        catch (SocketException)
        {
            // Already gone.
        }
    }

    public void Dispose()
    {
        Close();
        _socket.Dispose();
    }
}
